### PR TITLE
Oap use VectorizedReader read Parquet

### PR DESCRIPTION
**Pull request by LuciferYang**
## What changes were proposed in this pull request?
Task 6 of #536

1. Add `VectorizedContext` encapsulation infomation for Vectorized Read, `partitionColumns` and `partitionValues` use by `VectorizedOapRecordReader#initBatch` method, `returningBatch` use by `VectorizedOapRecordReader#enableReturningBatches` method.

2. Add `Option[VectorizedContext]` to `DataFile` Constructor and transfer it from `OapFileFormat`

3. `OapFileFormat` redefine supportBatch for Parquet File, this method use by `FileSourceScanExec` to init supportsBatch condition and buildReaderWithPartitionValues to know if support read data returningBatches.

4. `OapFileFormat` refer to `ParquetFileFormat`, use resultSchema to decide if this query support Vectorized Read and returningBatch and pass context to `OapDataReader`.

5. `OapFileFormat#buildReaderWithPartitionValues` method return iter directly if `enableVectorizedReader == true` because of partitionValues already filled by VectorizedReader, else use original branch.

6.  `ParquetDataFile` use `VectorizedOapRecordReader` or `IndexedVectorizedOapRecordReader` read Parquet file when `VectorizedContext` is not None.

7. Add `VectorizedDataSuite` to test ParquetDataFile read data full scan or Indexed. 

8. Add `batch.isFiltered($rowidx)` process to `FileSourceScanExec#doProduceVectorized` generate Java code to skip  write row to outbuffer by index data. 

## How was this patch tested?

1. existing sql case

2. VectorizedDataSuite

### CODE REVIEW

**Comment by LuciferYang** [comment]
Base on this pr, we will do comparison test between isFiltered is used by DataSourceScanExec or not  #598

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala [hunk 1]
```diff
@@ -126,8 +126,18 @@ private[sql] class OapFileFormat extends FileFormat
...
+    // TODO remove readerClassName after oap support batch return
+    val readerClassName = meta match {
+      case Some(m) =>
+        m.dataReaderClassName
+      case _ => ""
+    }
+    val conf = sparkSession.sessionState.conf
+    // TODO modify conditions after oap support batch return
+    readerClassName.equals(OapFileFormat.PARQUET_DATA_FILE_CLASSNAME) &&
+      conf.parquetVectorizedReaderEnabled && conf.wholeStageEnabled &&
```
Comment (xuanyuanking) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: new line for `conf.wholeStageEnable`

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala [hunk 2]
```diff
@@ -274,6 +284,16 @@ private[sql] class OapFileFormat extends FileFormat
         val requiredIds = requiredSchema.map(dataSchema.fields.indexOf(_)).toArray
         val pushed = FilterHelper.tryToPushFilters(sparkSession, requiredSchema, filters)
 
+        // refer to ParquetFileFormat
+        // use resultSchema to decide if
```
Comment (xuanyuanking) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: strange line break here

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala [hunk 3]
```diff
@@ -305,17 +325,32 @@ private[sql] class OapFileFormat extends FileFormat
...
               Option(TaskContext.get()).foreach(_.addTaskCompletionListener(_ => iter.close()))
               oapMetrics.updateIndexAndRowRead(reader, totalRows)
-
-              val fullSchema = requiredSchema.toAttributes ++ partitionSchema.toAttributes
-              val joinedRow = new JoinedRow()
-              val appendPartitionColumns = GenerateUnsafeProjection.generate(fullSchema, fullSchema)
-
-              iter.map(d => appendPartitionColumns(joinedRow(d, file.partitionValues)))
+              // if enableVectorizedReader == true, return iter directly
+              // else use original branch
```
Comment (xuanyuanking) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: also here.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala (hunk 1 above)
Comment (LuciferYang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: ok~

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala (hunk 2 above)
Comment (LuciferYang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: already rearrangement comments~

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala (hunk 3 above)
Comment (LuciferYang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: already rearrangement comments ~

**Comment by LuciferYang** [comment]
@xuanyuanking already re-format `supportBatch` method and rearrangement comments.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/OapDataFile.scala [hunk 4]
```diff
@@ -32,8 +32,11 @@ import org.apache.spark.sql.types._
 import org.apache.spark.util.CompletionIterator
 
 
-private[oap] case class OapDataFile(path: String, schema: StructType,
-                                    configuration: Configuration) extends DataFile {
+private[oap] case class OapDataFile(
+    path: String,
+    schema: StructType,
+    configuration: Configuration,
+    context: Option[VectorizedContext] = None) extends DataFile {
```
[the remark this answers is not in the document]
Comment (LuciferYang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/OapDataFile.scala]: OK, I'll add a method to set VectorizedContext in ParquetDataFile and revert this API change.

**Comment by LuciferYang** [comment]
@linhong-intel Add a method named setVectorizedContext in ParquetDataFile and avoid DataFile API change.

**Pull request merged**